comp/dogstatsd/listeners/uds_common.go

```diff
@@ -172,99 +189,116 @@ func (l *UDSListener) Listen() {
 			capBuff = replay.CapPool.Get().(*replay.CaptureBuffer)
 			capBuff.Pb.Ancillary = nil
 			capBuff.Pb.Payload = nil
+			capBuff.Pb.Pid = 0
+			capBuff.Pb.AncillarySize = int32(0)
+			capBuff.Pb.PayloadSize = int32(0)
 			capBuff.ContainerID = ""
 		}
 
 		if l.OriginDetection {
 			// Read datagram + credentials in ancillary data
-			oob := l.oobPoolManager.Get().(*[]byte)
-			oobS := *oob
-			var oobn int
+			oob = l.oobPoolManager.Get().(*[]byte)
+			oobS = *oob
+		}
 
-			if rateLimiter != nil {
-				if err = rateLimiter.MayWait(); err != nil {
-					log.Error(err)
-				}
+		if rateLimiter != nil {
+			if err = rateLimiter.MayWait(); err != nil {
+				log.Error(err)
 			}
+		}
 
-			t2 = time.Now()
-			tlmListener.Observe(float64(t2.Sub(t1).Nanoseconds()), "uds")
+		t2 = time.Now()
+		tlmListener.Observe(float64(t2.Sub(t1).Nanoseconds()), l.network, "uds")
+
+		var expectedPacketLength int32
+		var maxPacketLength int32
+		if l.network == "unix" {
+			// Read the expected packet length (in stream mode)
+			err = binary.Read(conn, binary.BigEndian, &expectedPacketLength)
+			switch {
+			case err == io.EOF, errors.Is(err, io.ErrUnexpectedEOF):
+				log.Debugf("dogstatsd-uds: %s connection closed", l.network)
+				return nil
+			}
+			maxPacketLength = expectedPacketLength
+		} else {
+			maxPacketLength = int32(len(packet.Buffer))
+		}
 
-			n, oobn, _, _, err = l.conn.ReadMsgUnix(packet.Buffer, oobS)
+		for err == nil {
+			if oob != nil {
+				n, oobn, _, _, err = conn.ReadMsgUnix(packet.Buffer[n:maxPacketLength], oobS[oobn:])
+			} else {
+				n, _, err = conn.ReadFromUnix(packet.Buffer[n:maxPacketLength])
+			}
+			if n == 0 && oobn == 0 {
+				log.Debugf("dogstatsd-uds: %s connection closed", l.network)
+				return nil
+			}
+			// If framing is disabled (unixgram, unixpacket), we always will have read the whole packet
+			if expectedPacketLength == 0 {
+				break
+			}
+			// Otherwise see if we need to continue to accumulate bytes or not
+			if int32(n) == expectedPacketLength {
+				break
+			}
+			if int32(n) > expectedPacketLength {
```

Comment:
Is this never suppose to happen? Like, you should never receive the start of a next "message"?

Reply:
no it's never supposed to happen, but I wanted to be a bit defensive here

Reply:
Makes sense, thanks.